opensubtitles/opensubtitles.go

```diff
@@ -0,0 +1,248 @@
+type Client struct {
+	clientMu sync.Mutex   // clientMu protects the client during calls that modify the CheckRedirect func.
+	client   *http.Client // HTTP client used to communicate with the API.
+
+	// Base URL for API requests.
+	BaseURL *url.URL
+
+	// Base URL for uploading files.
+	UploadURL *url.URL
+
+	// User agent used when communicating with the OpenSubtitles API.
+	UserAgent string
+
+	Credential Credentials
+
+	rateMu sync.Mutex
+
+	Token string
+	//rateLimits [categories]Rate // Rate limits for the client as determined by the most recent API calls.
+
+	common service // Reuse a single struct instead of allocating one for each service on the heap.
+
+	// Services used for talking to different parts of the OpenSubtitles API.
+	Authentication *AuthenticationService
+	Discover       *DiscoverService
+	Download       *DownloadService
+	Find           *FindService
+	Info           *InfoService
+	Search         *SearchService
+}
+
+type service struct {
+	client *Client
+}
+
+// addOptions adds the parameters in opts as URL query parameters to s. opts
+// must be a struct whose fields may contain "url" tags.
+func addOptions(s string, opts interface{}) (string, error) {
+	v := reflect.ValueOf(opts)
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return s, nil
+	}
+
+	u, err := url.Parse(s)
+	if err != nil {
+		return s, err
+	}
+
+	qs, err := query.Values(opts)
+	if err != nil {
+		return s, err
+	}
+
+	u.RawQuery = qs.Encode()
+	return u.String(), nil
+}
+
+// NewClient returns a new OpenSubtitles API client. If a nil httpClient is
+// provided, a new http.Client will be used. To use API methods which require
+// authentication, provide an http.Client that will perform the authentication
+// for you (such as that provided by the golang.org/x/oauth2 library).
+func NewClient(httpClient *http.Client, token string, cred Credentials) (c *Client) {
+	if httpClient == nil {
+		httpClient = &http.Client{}
+	}
+	baseURL, _ := url.Parse(defaultBaseURL)
+	//uploadURL, _ := url.Parse(uploadBaseURL)
+
+	c = &Client{client: httpClient, BaseURL: baseURL, UserAgent: userAgent, Token: token}
+	c.common.client = c
+	c.Authentication = (*AuthenticationService)(&c.common)
+	c.Discover = (*DiscoverService)(&c.common)
+	c.Download = (*DownloadService)(&c.common)
+	c.Find = (*FindService)(&c.common)
+	c.Info = (*InfoService)(&c.common)
+	c.Search = (*SearchService)(&c.common)
+
+	//Check if struct Credential is not empty
+	if (Credentials{}) != cred {
+		c.Credential = cred
+	}
+
+	return
+}
+//Connect return a new Client with a working token by making the authentication
+//with the Authentication Login function.
+func (c *Client) Connect() (*Client, error) {
+	if c == nil {
+		return nil, errors.New("no client provided")
+	}
+	if len(c.Token) < 1 {
+		log, resp, _ := c.Authentication.Login(context.Background(), &c.Credential)
+		if (&LoggedIn{}) != log {
+			if len(log.Token) > 0 && resp.StatusCode == http.StatusOK {
```

Comment:
Codacy found an issue: [extra empty line at the end of a block](https://app.codacy.com/manual/TheForgotten69/go-opensubtitles/pullRequest?prid=6093402)